### PR TITLE
Improve ign tool support on macOS

### PR DESCRIPTION
**Pull request by scpeters**
The `ign gazebo` command does not have full support on macOS (see #25 and #44). If System Integrity Protection is not disabled, the system version of ruby at `/usr/bin/ruby` cannot be used, though the brew version of ruby can be used as a workaround. Additionally, the GUI interface window does not work on macOS. This pull request enables `UNIT_ign_TEST` on macOS and fixes it to use the brew version of ruby (which was added as a dependency to the homebrew formulae in https://github.com/osrf/homebrew-simulation/pull/1221). It also moves the configuration of environment variables from the test to the cmake code.

I've also added some extra console error messages when macOS is detected (from the `.dylib` suffix on library files):

~~~
$ ign gazebo
Library error for [/Users/ignition/ws/install/ignition-gazebo4/lib/libignition-gazebo4-ign.4.0.0.dylib]: can't load /Users/ignition/ws/install/ignition-gazebo4/lib/libignition-gazebo4-ign.4.0.0.dylib

If this script was executed with /usr/bin/ruby, this error may be caused by
macOS System Integrity Protection. One workaround is to use a different
version of ruby, for example:
    brew install ruby
and add the following line to your shell profile:
    export PATH=/usr/local/opt/ruby/bin:$PATH
If you are using a colcon workspace, please ensure that the setup script
has properly set the DYLD_LIBRARY_PATH environment variables.
~~~

~~~
# explicitly invoking with brew's ruby and without -g or -s
$ /usr/local/opt/ruby/bin/ruby $(which ign) gazebo
`ign gazebo` currently only works with the -s argument on macOS.
~~~

~~~
# explicitly invoking with brew's ruby and with -g
$ /usr/local/opt/ruby/bin/ruby $(which ign) gazebo -g
`ign gazebo` currently only works with the -s argument on macOS.
~~~

### CODE REVIEW

**Comment by scpeters** [comment]
ign tests passing on macOS: https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-homebrew-amd64/4218/testReport/(root)/CmdLine/

File: CMakeLists.txt
Comment (chapulina) [inline CMakeLists.txt:127]: Can we make this required? This should work:

```
ign_find_package(ignition-tools
  REQUIRED
  PKGCONFIG "ignition-tools")
```

See https://github.com/ignitionrobotics/ign-tools/issues/23

File: src/cmd/cmdgazebo.rb.in
Comment (chapulina) [inline src/cmd/cmdgazebo.rb.in:450]: How about adding a comment with a link to https://github.com/ignitionrobotics/ign-gazebo/issues/44?

File: src/CMakeLists.txt
Comment (chapulina) [inline src/CMakeLists.txt:161]: Interesting, I don't think we've been setting environment variables for tests from CMake.

File: src/CMakeLists.txt
Comment (chapulina) [inline src/CMakeLists.txt:138]: This feels a bit out of place here just for the tests. It feels to me that it should either be something that we set at a higher level, such as:

* when we find Ruby using `ign-cmake`
* when we find Ruby for `ign-tools`, and export it to all its dependencies

Or it should be set by the user / machine running the tests, outside of the test code.

My main concerns are:

* this place may be a bit hidden for someone trying to understand howcome the tests pass but their setup fails
* having to duplicate this logic for all similar command line tests
* having a different environment for runtime and for testing

Any thoughts?

**Review by chapulina — COMMENTED** [review]
Great to give more direction to macOS users!

File: src/CMakeLists.txt
Comment (scpeters) [reply src/CMakeLists.txt:161]: I learned it from Dirk; it's not universal, but we have done it in a few places:

* https://github.com/ignitionrobotics/ign-cmake/blob/ignition-cmake2_2.6.0/examples/CMakeLists.txt#L180-L186
* https://github.com/osrf/sdformat/blob/sdformat10_10.0.0/cmake/SDFUtils.cmake#L192-L197
* https://github.com/osrf/gazebo/blob/gazebo11_11.3.0/cmake/GazeboTestUtils.cmake#L69-L77

I think it's a bit cleaner to set them from cmake. One drawback is that executing the binaries directly (like `./UNIT_ign_TEST`) may not work, but it's easy enough to use `ctest -R UNIT_ign_TEST -VV` once you know to do so

Comment (chapulina) [reply src/CMakeLists.txt:161]: >  One drawback is that executing the binaries directly (like ./UNIT_ign_TEST) may not work

Oh no, that sounds like a regression. We have that workflow documented in a few places, it would be unfortunate not to support it anymore.

File: src/CMakeLists.txt
Comment (scpeters) [reply src/CMakeLists.txt:138]: > This feels a bit out of place here just for the tests. It feels to me that it should either be something that we set at a higher level, such as:
> 
> * when we find Ruby using `ign-cmake`
> * when we find Ruby for `ign-tools`, and export it to all its dependencies

we don't actually find or export Ruby in `ign-tools`; whichever `ruby` is in the `PATH` at runtime will be used

the only place I know of where we find Ruby during build time is sdformat, but that's just to generate xml schema files, and it doesn't become a run-time dependency (except via `ign-tools`, but that is similar to our other packages)

> Or it should be set by the user / machine running the tests, outside of the test code.

this would require additional configuration of our CI machines / build scripts. I thought it would be simpler to give cmake a hint about which ruby to prefer

> * this place may be a bit hidden for someone trying to understand howcome the tests pass but their setup fails

I've added a console error message to `ign gazebo` specifically for macOS when there are problems that seem related to SIP that recommends using the brew version of ruby to mitigate this

> * having to duplicate this logic for all similar command line tests

in most of our other packages, the `ign.cc` files are linked directly into the main package library, but here it is linked into a component library. I believe SIP causes problems with the RPATH relating the component library to the main project library when trying to dlopen, which is why it is an issue here (and in https://github.com/ignitionrobotics/ign-plugin/pull/32 which uses the loader component) but not in ign-transport or ign-msgs. I don't know exactly how SIP works, but this is my best guess

> * having a different environment for runtime and for testing

there is a console message recommending the brew ruby if there are errors, which should encourage people to use a setup that matches our test environment

Comment (chapulina) [reply src/CMakeLists.txt:138]: > we don't actually find or export Ruby in ign-tools; whichever ruby is in the PATH at runtime will be used

Sure, but what I'm wondering is whether this is the best behaviour for users of `ign-tools`.

> in most of our other packages, the ign.cc files are linked directly into the main package library, but here it is linked into a component library.

I see, so the fix isn't needed for all libraries, so it probably makes sense to do it only for the libraries that need it, instead of forcing everything that depends on `ign-tools` to use ruby from brew.

---

Thanks for the clarifications, I think it's ok like this.

File: src/cmd/cmdgazebo.rb.in
Comment (scpeters) [reply src/cmd/cmdgazebo.rb.in]: added in 0dfa9aa175adf053c068ac087e8311b09e9897ad

File: CMakeLists.txt
Comment (scpeters) [reply CMakeLists.txt]: added in 0dfa9aa175adf053c068ac087e8311b09e9897ad

File: src/CMakeLists.txt
Comment (chapulina) [reply src/CMakeLists.txt:161]: As I mentioned during the meeting, I don't see the advantage of moving the environment variables to CMake. 

It prevents a common workflow, and it hides the environment configuration from anyone who's debugging the tests or looking into them to understand how the code works. Especially for variables like `IGN_GAZEBO_SYSTEM_PLUGIN_PATH`, which are set by the user at runtime, I think it's valuable to explicitly set them during the tests.

But I'm not going to block this PR on this, I'll leave it up to you.

**Review by chapulina — APPROVED** [review]

**Comment by chapulina** [comment]
@scpeters , I think you moved this back to "In Progress" because a resolution to https://github.com/ignitionrobotics/ign-tools/issues/7 may help here, right? I think we could get this in as is, and make those improvements later if we can. Let me know if I misunderstood something. I think this is a good intermediate state and better than the current state without it.

File: src/CMakeLists.txt
Comment (scpeters) [reply src/CMakeLists.txt:161]: I pointed to 3 examples in our code where it's already being used, so this isn't a completely new thing, but you're right that the docs need to be updated to handle this case. I'll merge this then open a pull request to update the docs

**Pull request merged**
